// Code generated by protoc-gen-validate. DO NOT EDIT.
// source: auth/permission_internal_service.proto

package auth

import (
	"bytes"
	"errors"
	"fmt"
	"net"
	"net/mail"
	"net/url"
	"regexp"
	"sort"
	"strings"
	"time"
	"unicode/utf8"

	"google.golang.org/protobuf/types/known/anypb"
)

// ensure the imports are used
var (
	_ = bytes.MinRead
	_ = errors.New("")
	_ = fmt.Print
	_ = utf8.UTFMax
	_ = (*regexp.Regexp)(nil)
	_ = (*strings.Reader)(nil)
	_ = net.IPv4len
	_ = time.Duration(0)
	_ = (*url.URL)(nil)
	_ = (*mail.Address)(nil)
	_ = anypb.Any{}
	_ = sort.Sort
)

// Validate checks the field values on PermissionOfUserResp with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *PermissionOfUserResp) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on PermissionOfUserResp with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// PermissionOfUserRespMultiError, or nil if none found.
func (m *PermissionOfUserResp) ValidateAll() error {
	return m.validate(true)
}

func (m *PermissionOfUserResp) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Success

	switch v := m.Resp.(type) {
	case *PermissionOfUserResp_Data_:
		if v == nil {
			err := PermissionOfUserRespValidationError{
				field:  "Resp",
				reason: "oneof value cannot be a typed-nil",
			}
			if !all {
				return err
			}
			errors = append(errors, err)
		}

		if all {
			switch v := interface{}(m.GetData()).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, PermissionOfUserRespValidationError{
						field:  "Data",
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, PermissionOfUserRespValidationError{
						field:  "Data",
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(m.GetData()).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return PermissionOfUserRespValidationError{
					field:  "Data",
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	case *PermissionOfUserResp_Error:
		if v == nil {
			err := PermissionOfUserRespValidationError{
				field:  "Resp",
				reason: "oneof value cannot be a typed-nil",
			}
			if !all {
				return err
			}
			errors = append(errors, err)
		}

		if all {
			switch v := interface{}(m.GetError()).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, PermissionOfUserRespValidationError{
						field:  "Error",
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, PermissionOfUserRespValidationError{
						field:  "Error",
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(m.GetError()).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return PermissionOfUserRespValidationError{
					field:  "Error",
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	default:
		_ = v // ensures v is used
	}

	if len(errors) > 0 {
		return PermissionOfUserRespMultiError(errors)
	}

	return nil
}

// PermissionOfUserRespMultiError is an error wrapping multiple validation
// errors returned by PermissionOfUserResp.ValidateAll() if the designated
// constraints aren't met.
type PermissionOfUserRespMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m PermissionOfUserRespMultiError) Error() string {
	msgs := make([]string, 0, len(m))
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m PermissionOfUserRespMultiError) AllErrors() []error { return m }

// PermissionOfUserRespValidationError is the validation error returned by
// PermissionOfUserResp.Validate if the designated constraints aren't met.
type PermissionOfUserRespValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e PermissionOfUserRespValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e PermissionOfUserRespValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e PermissionOfUserRespValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e PermissionOfUserRespValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e PermissionOfUserRespValidationError) ErrorName() string {
	return "PermissionOfUserRespValidationError"
}

// Error satisfies the builtin error interface
func (e PermissionOfUserRespValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sPermissionOfUserResp.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = PermissionOfUserRespValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = PermissionOfUserRespValidationError{}

// Validate checks the field values on PermissionOfUserResp_Data with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *PermissionOfUserResp_Data) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on PermissionOfUserResp_Data with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// PermissionOfUserResp_DataMultiError, or nil if none found.
func (m *PermissionOfUserResp_Data) ValidateAll() error {
	return m.validate(true)
}

func (m *PermissionOfUserResp_Data) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	for idx, item := range m.GetPermissions() {
		_, _ = idx, item

		if all {
			switch v := interface{}(item).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, PermissionOfUserResp_DataValidationError{
						field:  fmt.Sprintf("Permissions[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, PermissionOfUserResp_DataValidationError{
						field:  fmt.Sprintf("Permissions[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(item).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return PermissionOfUserResp_DataValidationError{
					field:  fmt.Sprintf("Permissions[%v]", idx),
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	if len(errors) > 0 {
		return PermissionOfUserResp_DataMultiError(errors)
	}

	return nil
}

// PermissionOfUserResp_DataMultiError is an error wrapping multiple validation
// errors returned by PermissionOfUserResp_Data.ValidateAll() if the
// designated constraints aren't met.
type PermissionOfUserResp_DataMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m PermissionOfUserResp_DataMultiError) Error() string {
	msgs := make([]string, 0, len(m))
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m PermissionOfUserResp_DataMultiError) AllErrors() []error { return m }

// PermissionOfUserResp_DataValidationError is the validation error returned by
// PermissionOfUserResp_Data.Validate if the designated constraints aren't met.
type PermissionOfUserResp_DataValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e PermissionOfUserResp_DataValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e PermissionOfUserResp_DataValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e PermissionOfUserResp_DataValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e PermissionOfUserResp_DataValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e PermissionOfUserResp_DataValidationError) ErrorName() string {
	return "PermissionOfUserResp_DataValidationError"
}

// Error satisfies the builtin error interface
func (e PermissionOfUserResp_DataValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sPermissionOfUserResp_Data.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = PermissionOfUserResp_DataValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = PermissionOfUserResp_DataValidationError{}
